c10/test/util/complex_test_common.h

```diff
@@ -0,0 +1,464 @@
+  static_assert(alignof(c10::complex<double>) == 16, "");
+}
+
+MAYBE_GLOBAL void test_pod() {
+  static_assert(std::is_standard_layout<c10::complex<float>>::value, "");
+  static_assert(std::is_standard_layout<c10::complex<double>>::value, "");
+}
+
+TEST(TestMemory, ReinterpretCast) {
+  std::complex<float> z(1, 2);
+  c10::complex<float> zz = *reinterpret_cast<c10::complex<float>*>(&z);
+  ASSERT_EQ(zz.real(), float(1));
+  ASSERT_EQ(zz.imag(), float(2));
+
+  std::complex<double> zzz(1, 2);
+  c10::complex<double> zzzz = *reinterpret_cast<c10::complex<double>*>(&zzz);
+  ASSERT_EQ(zzzz.real(), double(1));
+  ASSERT_EQ(zzzz.imag(), double(2));
+}
+
+}  // memory
+
+namespace constructors {
+
+template<typename scalar_t>
+C10_HOST_DEVICE void test_construct_from_scalar() {
+  constexpr scalar_t num1 = scalar_t(1.23);
+  constexpr scalar_t num2 = scalar_t(4.56);
+  constexpr scalar_t zero = scalar_t();
+  static_assert(c10::complex<scalar_t>(num1, num2).real() == num1, "");
+  static_assert(c10::complex<scalar_t>(num1, num2).imag() == num2, "");
+  static_assert(c10::complex<scalar_t>(num1).real() == num1, "");
+  static_assert(c10::complex<scalar_t>(num1).imag() == zero, "");
+  static_assert(c10::complex<scalar_t>().real() == zero, "");
+  static_assert(c10::complex<scalar_t>().imag() == zero, "");
+}
+
+template<typename scalar_t, typename other_t>
+C10_HOST_DEVICE void test_construct_from_other() {
+  constexpr other_t num1 = other_t(1.23);
+  constexpr other_t num2 = other_t(4.56);
+  constexpr scalar_t num3 = scalar_t(num1);
+  constexpr scalar_t num4 = scalar_t(num2);
+  static_assert(c10::complex<scalar_t>(c10::complex<other_t>(num1, num2)).real() == num3, "");
+  static_assert(c10::complex<scalar_t>(c10::complex<other_t>(num1, num2)).imag() == num4, "");
+}
+
+MAYBE_GLOBAL void test_convert_constructors() {
+  test_construct_from_scalar<float>();
+  test_construct_from_scalar<double>();
+
+  static_assert(std::is_convertible<c10::complex<float>, c10::complex<float>>::value, "");
+  static_assert(!std::is_convertible<c10::complex<double>, c10::complex<float>>::value, "");
+  static_assert(std::is_convertible<c10::complex<float>, c10::complex<double>>::value, "");
+  static_assert(std::is_convertible<c10::complex<double>, c10::complex<double>>::value, "");
+
+  static_assert(std::is_constructible<c10::complex<float>, c10::complex<float>>::value, "");
+  static_assert(std::is_constructible<c10::complex<double>, c10::complex<float>>::value, "");
+  static_assert(std::is_constructible<c10::complex<float>, c10::complex<double>>::value, "");
+  static_assert(std::is_constructible<c10::complex<double>, c10::complex<double>>::value, "");
+
+  test_construct_from_other<float, float>();
+  test_construct_from_other<float, double>();
+  test_construct_from_other<double, float>();
+  test_construct_from_other<double, double>();
+}
+
+template<typename scalar_t>
+C10_HOST_DEVICE void test_construct_from_std() {
```

Comment:
I don't think it should be C10_HOST_DEVICE because we are only testing construction of c10::complex from std::complex ??

Reply:
Construction from `std::complex` is `constexpr`, therefore it must be available on device too. A horrible trick here is: Although `std::complex` is generally not available on CUDA, `std::complex::real` and `std::complex::imag` is `constexpr`, and this trick is already being used on device to convert from `std::complex` to `thrust::complex`.